src/request.cc

```diff
@@ -0,0 +1,306 @@
+// Modeled from parser at: http://www.boost.org/doc/libs/1_62_0/doc/html/boost_asio/example/cpp11/http/server/request_parser.hpp
+// Implementation to parse HTTP requests
+
+#include "request.h"
+
+// Parse some data. The enum return value is good when a complete request has
+// been parsed, bad if the data is invalid, indeterminate when more data is
+// required. The InputIterator return value indicates how much of the input
+// has been consumed.
+std::unique_ptr<Request> Request::Parse(const std::string& raw_request) {
+  std::unique_ptr<Request> parsed_request(new Request());
+  for (char c : raw_request) {
+    parsed_request->parsed_status_ = parsed_request->consume(c);
```

Comment:
This is kind of ugly but the other two options are:

1. make a wrapper method that sets parsed_status_ = parsed_request->consume(c);
    (so the wrapper method would call consume and set status)

2. make consume() return void, with it setting parsed_status_ and also setting return; (cuz it's a switch statement)